File: challenger-backend/src/main/java/com/kameo/challenger/odb/ChallengeParticipantODB.java
package com.kameo.challenger.odb;

import com.kameo.challenger.odb.api.IIdentity;
import com.kameo.challenger.utils.odb.EntityHelper;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Date;

@Data
@Entity
@NoArgsConstructor
@Table(indexes = {
    @Index(columnList = "user_id"),
    @Index(columnList = "challenge_id"),
})
@ToString(of = "id")
public
class ChallengeParticipantODB implements IIdentity {
    @Id
    @GeneratedValue(strategy= GenerationType.AUTO)
    long id;


    @NotNull
    @ManyToOne
    ChallengeODB challenge;

    @NotNull
    @ManyToOne
    UserODB user;

    @Temporal(TemporalType.TIMESTAMP)
    Date lastSeen;


    @NotNull
    @Enumerated
    ChallengeStatus challengeStatus;

    public int hashCode() {
        return Long.hashCode(id);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof ChallengeParticipantODB && ((ChallengeParticipantODB) obj).getId()==this.getId();
    }

    public ChallengeODB getChallenge() {
        return challenge;
    }

    public void setChallenge(ChallengeODB challenge) {
        this.challenge = challenge;
    }

    public UserODB getUser() {
        return user;
    }

    public void setUser(UserODB user) {
        this.user = user;
    }

    public Date getLastSeen() {
        return lastSeen;
    }

    public void setLastSeen(Date lastSeen) {
        this.lastSeen = lastSeen;
    }


    @Override
    public boolean isNew() {
        return getId()<0;
    }
}

File: challenger-backend/src/main/java/com/kameo/challenger/domain/challenges/ChallengeParticipantODB.kt
package com.kameo.challenger.domain.challenges

import com.kameo.challenger.domain.challenges.ChallengeODB
import com.kameo.challenger.domain.challenges.ChallengeStatus
import com.kameo.challenger.odb.UserODB
import com.kameo.challenger.odb.api.IIdentity
import java.util.*
import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
data class ChallengeParticipantODB( @Id
                    @GeneratedValue(strategy= javax.persistence.GenerationType.AUTO)
                    override var id:Long=0) : IIdentity {


    @ManyToOne
    lateinit var challenge: ChallengeODB

    @ManyToOne
    lateinit var user: UserODB

    @Temporal(TemporalType.TIMESTAMP)
    var lastSeen: Date?=null;


    @NotNull
    @Enumerated
    var challengeStatus: ChallengeStatus = ChallengeStatus.WAITING_FOR_ACCEPTANCE;
}